agent/lib/kontena/watchdog.rb

```diff
@@ -0,0 +1,71 @@
+# The watchdog is intended to recover the agent from unrepairable errors by killing it, and allowing it to be restarted.
+#
+# The watchdog takes a block, which gets run in a separate thread every interval.
+# The watchdog will abort if the block times out or fails with an exception.
+#
+# The watchdog abort will immediately exit! the process, without running any at_exit hooks for shutdown.
+class Kontena::Watchdog
+  include Celluloid
+  include Kontena::Logging
+
+  INTERVAL = 10.0
+  TIMEOUT = 30.0
+
+  def self.watch(**options, &block)
+    # pass block as a proc, instead of as a celluloid block proxy
+    new(block, **options)
+  end
+
+  def initialize(block, interval: INTERVAL, timeout: TIMEOUT, abort_exit: true)
+    @block = block
+
+    @interval = interval
+    @timeout = timeout
+    @abort_exit = abort_exit
+
+    async.start
+  end
+
+  def start
+    info "watchdog start"
+    @timer = every(@interval) do
+      ping
+    end
+  end
+
+  # Stop the every loop from start
+  def stop
+    @timer.cancel
+  end
+
+  # Start new watchdog ping
+  def ping
+    start = Time.now
+
+    # use exclusive mode to prevent the task from suspending and causing the Timeout::Error to crash the actor thread instead
+    exclusive {
+      Timeout.timeout(@timeout) do
```

Comment:
BTW: using `Timeout.timeout` in a celluloid actor needs the `exclusive` mode, otherwise the block might cause the celluloid task fiber to suspend, and then the `Timeout::Error` will just abort the actor thread `run` loop and cause the `Watchdog` to crash without aborting:

```
E, [2017-12-13T16:20:37.495949 #1] ERROR -- : Actor crashed!
Timeout::Error: execution expired
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `sleep'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `block in check'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:33:in `while_time_remaining'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:11:in `for'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:58:in `check'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:155:in `block in run'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/group.rb:66:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:152:in `run'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:131:in `block in start'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-essentials-0.20.5/lib/celluloid/internals/thread_handle.rb:14:in `block in initialize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor/system.rb:78:in `block in get_thread'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:50:in `block in instantiate'
```